src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala

```diff
@@ -160,11 +160,13 @@ trait MatchCodeGen extends Interface {
         // returns MatchMonad[U]
         def flatMap(prev: Tree, b: Symbol, next: Tree): Tree = {
           val prevSym = freshSym(prev.pos, prev.tpe, "o")
+          // avoid a constant type on isEmpty being constant folded when the tree is typed
+          val isEmptySym = prev.tpe.member(vpmName.isEmpty).setInfo(NullaryMethodType(BooleanTpe))
           BLOCK(
             ValDef(prevSym, prev),
             // must be isEmpty and get as we don't control the target of the call (prev is an extractor call)
             ifThenElseZero(
-              NOT(prevSym DOT vpmName.isEmpty),
+              NOT((prevSym DOT vpmName.isEmpty).setSymbol(isEmptySym)),
```

Comment:
@lrytz @retronym does this look reasonable to you?  The relevant adaptation line is https://github.com/scala/scala/blob/f7d0e1fd8768cfc0d7fa7237f900e68d6d31d92a/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L1249 and I was using [`882972d` (#7620)](https://github.com/scala/scala/pull/7620/commits/882972d3a850887210b5c9041ec9b4d9557cfc2e) as a reference, but I couldn't find a way to do something equivalent to the `deriveCaseDef`/`deconst` it does there, so I ended up ignoring details and doing this.